keps/sig-apps/4212-declarative-node-maintenance/README.md

```diff
@@ -0,0 +1,2168 @@
+  node shutdown, the shutdown might not be registered again, and pods might be terminated
+  ungracefully. Also, new workloads can get scheduled on the node while the node is shutting down.
+  Cluster admins should, therefore, plan the maintenance in advance and ensure that pods are
+  gracefully removed before attempting to shut down or restart the machine.
+- The kubelet has no way of reliably detecting ongoing maintenance if the node is restarted in the
+  meantime.
+- Graceful termination of static pods during a shutdown is not possible today. It is also not
+  currently possible to prevent them from starting back up immediately after the machine has been
+  restarted and the kubelet has started again, if the node is still under maintenance. 
+
+### Motivation Summary
+
+To sum up. Some applications solve the disruption problem by introducing validating admission
+webhooks. This has some drawbacks. The webhooks are not easily discoverable by cluster admins. And
+they can block evictions for other applications if they are misconfigured or misbehave. The
+eviction API is not intended to be extensible in this way. The webhook approach is therefore not
+recommended.
+
+Some drainers solve the node drain by depending on the kubectl logic, or by extending/rewriting it
+with additional rules and logic.
+
+As seen in the experience reports and GitHub issues, some admins solve their problems by simply
+ignoring PDBs which can cause unnecessary disruptions or data loss. Some solve this by playing
+with the application deployment, but have to understand that the application supports this.
+
+kubelet's Graceful Node Shutdown feature is a best-effort solution for unplanned shutdowns, but
+it is not sufficient to ensure application and data safety.
+
+### Goals
+- Introduce NodeMaintenance API.
+- Introduce a node maintenance controller that creates evacuations.
+- Deprecate kubectl drain in favor of NodeMaintenance.
+- Implement NodeMaintenanceAwareKubelet feature to implement a lifecycle for static pods during a
+  maintenance.
+- Implement NodeMaintenanceAwareDaemonSet feature to prevent the scheduling of DaemonSet pods on
+  nodes during a maintenance.
+
+### Non-Goals
+- Introduce a node maintenance period, nodeDrainTimeout (similar to [cluster-api](https://cluster-api.sigs.k8s.io/developer/architecture/controllers/control-plane)
+  nodeDrainTimeout) or a TTL optional field as an upper bound on the duration of node maintenance.
+  Then the node maintenance would be garbage collected and the node made schedulable again.
+- Solve the node lifecycle management or automatic shutdown after the node drain is completed.
+  Implementation of this is better suited for other cluster components and actors who can use the
+  node maintenance as a building block to achieve their desired goals.
+- Synchronize all pod termination mechanisms (see #8 in the [Motivation](#motivation) section), so that they do
```

Comment:
This require to synchronize the Endpoints/Services too , after a quick skim to the KEP I wonder if this approach will not hit the same problem @smarterclayton is describing here https://github.com/kubernetes/kubernetes/issues/116965#issuecomment-2113227772, the Pods should report the status to avoid traffic to be sent to them if they are not ready

Reply:
As this is a non goal I think the KEP can comfortably find an easy way to not deliver it.

Reply:
1. This KEP does not propose to solve the non-goals nor does it suggest how or whether to implement them. It is a simply a list of things that could be done.
2. Even the goals section expects additional KEPs on the node and apps side to fully examine the problem. But for the sake of simplicity, I keep the NodeMaintenanceAwareDaemonSet and NodeMaintenanceAwareKubelet features here for now.

Regarding what @smarterclayton wrote, I think we have a lot of room to solve that. By utilizing both the NodeMaintenance and Evacuation API and deciding which parts should be mandatory and which optional.